src/combinatorics/set_family/shadow.lean

```diff
@@ -0,0 +1,131 @@
+/-
+Copyright (c) 2021 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta, Alena Gusakov, Yaël Dillies
+-/
+import data.finset.lattice
+
+/-!
+# Shadows
+
+This file defines shadows of a set family. The shadow of a set family is the set family of sets we
+get by removing any element from any set of the original family. If one pictures `finset α` as a big
+hypercube (each dimension being membership of a given element), then taking the shadow corresponds
+to projecting each finset down once in all available directions.
+
+## Main definitions
+
+The `shadow` of a set family is everything we can get by removing an element from each set.
+
+## Notation
+
+`∂ 𝒜` is notation for `shadow 𝒜`. It is situated in locale `finset_family`.
+
+We also maintain the convention that `a, b : α` are elements of the ground type, `A, B : finset α`
+are finsets, and `𝒜, ℬ : finset (finset α)` are finset families.
+
+## References
+
+* https://github.com/b-mehta/maths-notes/blob/master/iii/mich/combinatorics.pdf
+* http://discretemath.imp.fu-berlin.de/DMII-2015-16/kruskal.pdf
+
+## Tags
+
+shadow, set family
+-/
+
+open finset nat
+
+variables {α : Type*}
+
+/-!
+### Shadows
+
+The shadow of a set family is everything we can get by removing an element
+from each set.
+
+This section develops the introductory theory of shadows, with some lemmas on
+iterated shadows as well.
+-/
+
+variables [decidable_eq α] {𝒜 : finset (finset α)} {A B : finset α} {a : α}
+
+/-- The shadow of a set family `𝒜` is all sets we can get by removing one element
+from any set in `𝒜`, and the (`k` times) iterated shadow is all sets we can get
+by removing k elements from any set in `𝒜`. -/
+def shadow (𝒜 : finset (finset α)) : finset (finset α) := 𝒜.sup (λ A, A.image (erase A))
```

Comment:
I think this probably belongs in the `finset` namespace since you could define it on sets of sets too.

Reply:
An opinion, @b-mehta?

Reply:
Another argument would be dot notation, so you can do `𝒜.shadow`

Reply:
I'm with Eric on this - while the shadow is really only used in the context of finite objects, it can be defined sensibly on sets of sets too. Plus in context we'll mostly use the `∂` notation so it shouldn't cause many changes in this file (or subsequent ones)

Reply:
Here's what looks like two equivalent sensible definitions of the `set` version:
```lean
def set.shadow {α} (𝒜 : set (set α)) : set (set α) := ⋃ A ∈ 𝒜, (λ y, A \ {y}) '' A
def set.shadow {α} (𝒜 : set (set α)) : set (set α) := ⋃ A ∈ 𝒜, (((\) A) ∘ singleton) '' A
```
There's no need to include it in this PR, but I don't think it's unimaginable that someone might want to use this and `set.finite` instead; so let's play it safe with namespacing.

Reply:
But then I guess I should namespace *everything*? One doesn't really put in a namespace a definition but not its lemmas.